pyro/contrib/oed/eig.py

```diff
@@ -68,61 +73,201 @@ def posterior_entropy(y_dist, design):
+    retrace.compute_log_prob()
+    marginal_lp = logsumexp(sum(retrace.nodes[l]["log_prob"] for l in observation_labels), 0) \
+        - np.log(M)
 
     return (conditional_lp - marginal_lp).sum(0)/N
 
 
-def donsker_varadhan_loss(model, observation_label, T):
+def donsker_varadhan_eig(model, design, observation_labels, target_labels,
+                         num_samples, num_steps, T, optim, return_history=False,
+                         final_design=None, final_num_samples=None):
     """
     Donsker-Varadhan estimate of the expected information gain (EIG).
 
     The Donsker-Varadhan representation of EIG is
 
-        :math:`\\sup_T E[T(y, \\theta)] - \\log E[\\exp(T(\\bar{y}, \\bar{\\theta}))]`
+    .. math::
+
+        \\sup_T E_{p(y, \\theta | d)}[T(y, \\theta)] - \\log E_{p(y|d)p(\\theta)}[\\exp(T(\\bar{y}, \\bar{\\theta}))]
+
+    where :math:`T` is any (measurable) function.
 
-    where the first expectation is over the joint :math:`p(y | \\theta, d)` and
-    the second is over :math:`p(\\bar{y}|d)p(\\bar{\\theta})``.
+    This methods optimises the loss function over a pre-specified class of
+    functions `T`.
 
-    :param function model: A stochastic function.
-    :param str observation_label: String label for observed variable.
+    :param function model: A pyro model accepting `design` as only argument.
+    :param torch.Tensor design: Tensor representation of design
+    :param list observation_labels: A subset of the sample sites
+        present in `model`. These sites are regarded as future observations
+        and other sites are regarded as latent variables over which a
+        posterior is to be inferred.
+    :param list target_labels: A subset of the sample sites over which the posterior
+        entropy is to be measured.
+    :param int num_samples: Number of samples per iteration.
+    :param int num_steps: Number of optimisation steps.
     :param function or torch.nn.Module T: optimisable function `T` for use in the
         Donsker-Varadhan loss function.
+    :param pyro.optim.Optim optim: Optimiser to use.
+    :param bool return_history: If `True`, also returns a tensor giving the loss function
+        at each step of the optimisation.
+    :param torch.Tensor final_design: The final design tensor to evaluate at. If `None`, uses
+        `design`.
+    :param int final_num_samples: The number of samples to use at the final evaluation, If `None,
+        uses `num_samples`.
+    :return: EIG estimate, optionally includes full optimisatio history
+    :rtype: `torch.Tensor` or `tuple`
     """
+    if isinstance(observation_labels, str):
+        observation_labels = [observation_labels]
+    if isinstance(target_labels, str):
+        target_labels = [target_labels]
+    loss = donsker_varadhan_loss(model, T, observation_labels, target_labels)
+    return opt_eig_ape_loss(design, loss, num_samples, num_steps, optim, return_history,
```

Comment:
Just calls the `opt_eig_ape_loss` function with the right loss. This allows re-use of `opt` between DV and BA